Celeste.Mod.mm/Mod/Everest/Everest.Loader.cs

```diff
@@ -176,28 +176,35 @@ public static class Loader {
 
                 enforceOptionalDependencies = true;
 
-                string[] files = Directory.GetFiles(PathMods);
-                Array.Sort(files); //Prevent inode loading jank
-                for (int i = 0; i < files.Length; i++) {
-                    string file = Path.GetFileName(files[i]);
-                    if (!file.EndsWith(".zip") || !ShouldLoadFile(file))
-                        continue;
+                string[] files = Directory
```

Comment:
Did you profile these LINQ refactors? LINQ tends to allocate a bunch more objects than loop-based approaches, including intermediate enumerators and an entire second array at the end here, which you then also iterate through a second time because the loading now happens in a secondary pass after filtering.

Reply:
To obtain the accurate total number of mods to load, we need to iterate it at first before loading the files. I don't think this can be a performance bottleneck for just preprocessing an array of strings.

Also LINQ is even somehow faster than the for method.

![image](https://github.com/EverestAPI/Everest/assets/23134847/80e81da6-6e7f-46d0-a5b2-89bd3d4de168)

<details>
  <summary>The test code</summary>

```
using System;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using Microsoft.Diagnostics.Runtime;


namespace MyBenchmarks
{

    public class LINQVsFor
    {
        private const int N = 10000;
        private readonly byte[] data;

        private string[] Whitelist = Array.Empty<string>(), Blacklist = Array.Empty<string>(), TemporaryBlacklist = Array.Empty<string>();

        private bool ShouldLoadFile(string file)
        {
            return Whitelist != null ? Whitelist.Contains(file) : (!Blacklist.Contains(file) && (TemporaryBlacklist == null || !TemporaryBlacklist.Contains(file)));
        }

        private const string PathMods = "D:\\SteamLibrary\\steamapps\\common\\Celeste\\Mods";
        private string[] _files;
        private string tmp;


        public LINQVsFor()
        {
            _files = Directory.GetFiles(PathMods);
        }

        private void LoadMod(string file)
        {
            tmp = file;
        }

        [Benchmark]
        public void LINQ()
        {
            string[] files = _files
                    .OrderBy(f => f)
                    .Select(Path.GetFileName)
                    .Where(file => file.EndsWith(".zip") && ShouldLoadFile(file))
                    .ToArray();
            int len = files.Length;
            foreach (string file in files)
            {
                LoadMod(Path.Combine(PathMods, file));
            }
        }


        [Benchmark]
        public void For()
        {
            string[] files = _files;
            Array.Sort(files);
            int len = 0;
            for (int i = 0; i < files.Length; i++)
            {
                string file = Path.GetFileName(files[i]);
                if (!file.EndsWith(".zip") || !ShouldLoadFile(file))
                    continue;
                len++;
            }
            for (int i = 0; i < files.Length; i++)
            {
                string file = Path.GetFileName(files[i]);
                if (!file.EndsWith(".zip") || !ShouldLoadFile(file))
                    continue;
            }
        }


        [Benchmark]
        public void For2()
        {
            string[] files = _files;
            bool[] shouldLoad = new bool[files.Length];
            Array.Sort(files);
            int len = 0;
            for (int i = 0; i < files.Length; i++)
            {
                string file = Path.GetFileName(files[i]);
                shouldLoad[i] = file.EndsWith(".zip") && ShouldLoadFile(file);
                if (shouldLoad[i])
                    len++;
            }
            for (int i = 0; i < files.Length; i++)
            {
                if (shouldLoad[i])
                    LoadMod(files[i]);
            }
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<LINQVsFor>();
        }
    }
}
```
</details>

Reply:
Note: You should be careful with Linq microbenchmarks on modern .NET, as Profile Guided Optimisation can make Linq seem much much better in benchmarks than in real code - the JIT will just emit a fastpath directly into linq methods which inline everything, which it can't do in real apps due to how many distinct arguments linq methods receive.

Either way, perf here is realistically not a major concern, there are much bigger issues in loading than this :p